include/secp256k1_ecdsa_adaptor.h

```diff
@@ -10,7 +10,24 @@ extern "C" {
  *  Lloyd Fournier
  *  (https://lists.linuxfoundation.org/pipermail/lightning-dev/2019-November/002316.html
  *  and https://github.com/LLFourn/one-time-VES/blob/master/main.pdf).
-*/
+ *
+ *  WARNING! DANGER AHEAD!
+ *  As mentioned in Lloyd Fournier's paper, the adaptor signature leaks the
+ *  Diffie-Hellmann key between the signing key and the encryption key. This is
+ *  not a problem for ECDSA adaptor signatures themselves, but may result in a
+ *  complete loss of security when they are composed with other schemes. More
+ *  specifically, let us refer to the signer's public key as X = x*G, and to the
+ *  encryption key as Y = y*G. Given X, Y and the adaptor signature, it is
+ *  trivial to compute Y^x = X^y.
+ *  To completely avoid this problem, the signer can require a proof of
+ *  knowledge of the decryption key y before creating an adaptor signature. This
```

Comment:
Maybe explain how this proof of knowledge can be implemented.

On a second thought, I'm not sure how. It's formally not known to be a proof of knowledge but is an ECDSA signature enough here? Maybe recommend a Schnorr signature to be on the safe side? We have it in the library now, so why not.

Reply:
As far as I can tell, there are at least a few pitfalls with a Schnorr signature proof of knowledge. Firstly, a naive construction is replayable. For example, most of the schemes I've seen essentially just sign the generator (`c = H(g, y, t)` from https://en.wikipedia.org/wiki/Fiat%E2%80%93Shamir_heuristic and `c = H(g, h, u)` from https://crypto.stanford.edu/cs355/19sp/lec5.pdf). This can perhaps be resolved by requiring the verifier to send some random data to the prover to be used as the message. However, this opens a second potential attack, which is to trick the prover into signing a message, for example by sending "random data" that is in fact a transaction or some other non-random value. This second issue could be potentially resolved by having the prover hash the random data (and the prover should also probably include the nonce pubkey in the hash so that there's an input to the hash function which is not controlled by the verifier). So I think a scheme where (1) the verifier sends the prover some random data and (2) the prover hashes the random data along with the nonce pubkey and signs the hash, could be okay, but curious to hear what other folks think of the scheme.

Also, maybe suggesting a proof of knowledge scheme here is more trouble than it's worth because if the recipient of the adaptor signature knows the decryption key, then they can simply decrypt the adaptor signature when they receive it which seems like it would obviate the usefulness of encrypting the signature.

cc @jonasnick @real-or-random @LLFourn

Reply:
Hm, I hadn't considered that the recipient doesn't actually know the decryption key. Perhaps it would be okay in practice if before creating an adaptor signature for `enckey`, the signer would require a a signature for `enckey` and a message like "this is an adaptor sig enckey". Thus, whoever creates `enckey`s must also provide such a signature and thereby communicates that they don't care about the DH key between the signing and the `enckey`. The idea is that keys whose DH key with the signing key must not be leaked (such as an ephemeral ElGamal key) will not be used to create such signatures. However, this approach does not work for multi-hop-locks because there is no party that on its own knows the DL of the enckey when exchanging adaptor sigs to establish the route.

Reply:
@jesseposner I think the canonical way would be a signature on `H(m)` where `m` is a random challenge message  and `H` is a tagged hash with some unique tag such as `ECDSA adaptor PoK`. 

But I agree, the proof of knowledge seems not useful in this setting. 

Note that I also didn't notice this because my comments here are pretty superficial. You guys have a better understanding of the entire thing. 

Maybe warning like this is just fine: "A weaker defense is to simply not compose ECDSA adaptor signatures with other protocols."

Could we make that more specific?  "A defense is to not reuse the signing key of ECDSA adaptor signatures in other protocols." or "in the protocols that rely on the hardness of the CDH problem, e.g., Diffie-Hellman key exchange."

Reply:
I updated the [warning description](https://github.com/ElementsProject/secp256k1-zkp/pull/117/files#diff-0bc5e1a03ce026e8fea9bfb91a5334cc545fbd7ba78ad83ae5489b52e4e48856R14).

Reply:
> Also, maybe suggesting a proof of knowledge scheme here is more trouble than it's worth because if the recipient of the adaptor signature knows the decryption key, then they can simply decrypt the adaptor signature when they receive it which seems like it would obviate the usefulness of encrypting the signature.

Note the point is that the encryption is "one-time" in that it leaks the decryption key. The point of giving an adaptor signature to someone who knows the decryption key is precisely because you hope they will leak it by decrypting and publishing the signature. DLCs are an exception to this.

There was another attempt at improving terminology here by generalizing this to "lockable" signatures (still in my list of things to read):  https://eprint.iacr.org/2020/1613.pdf. Perhaps a "locked" signature is more intuitive than an "encrypted" signature. We already use the term "lock" in the sense of multi-hop locks so this might make things easier for people. Impressively they claim to be able to do it with BLS which I had thought was impossible to do efficiently.

> However, this approach does not work for multi-hop-locks because there is no party that on its own knows the DL of the enckey when exchanging adaptor sigs to establish the route.

Yep. Nothing much you can do here other than avoid CDH or direct people to using Schnorr on signet until it arrives on mainnet later this year 🤞 .

Reply:
>Note the point is that the encryption is "one-time" in that it leaks the decryption key. The point of giving an adaptor signature to someone who knows the decryption key is precisely because you hope they will leak it by decrypting and publishing the signature. DLCs are an exception to this.

Great point, for example, the last hop in a multi-hop payment will know the decryption key for the adaptor signature it receives. However, each hop before it will not know the decryption key when they receive their adaptor signatures (but they will of course learn the respective decryption keys during the settlement phase of the protocol).

>Perhaps a "locked" signature is more intuitive than an "encrypted" signature.

"Lock" is nice particularly because, as you point out, it makes the phrase "multi-hop locks" more clear, but then you need to refer to a "locking key" and an "unlocking key" and the "one-time" analogy to the OTP is lost, as is the scheme's verifiability. "One-time verifiably encrypted signatures" conveys more information to the reader about the scheme than "lockable signatures."